cpp/src_prims/sparse/coo.h

```diff
@@ -807,5 +807,163 @@ void coo_symmetrize(COO<T> *const in, COO<T> *out,
+  if (j % 2)
+    atomicAdd(&row_sizes[col], 1);
+  else
+    atomicAdd(&row_sizes2[col], 1);
+}
+
+/**
+ * @brief Reduce sum(row_sizes) + k
+ * Reduction for symmetric_find_size kernel. Allows algo to be faster.
+ *
+ * @param n: Number of rows
+ * @param k: Number of n_neighbors
+ * @param row_sizes: Input row sum 1 array(n)
+ * @param row_sizes2: Input row sum 2 array(n) for faster reduction
+ */
+__global__ static void reduce_find_size(const int n, const int k,
+                                        int *restrict row_sizes,
+                                        const int *restrict row_sizes2) {
+  const int i = (blockIdx.x * blockDim.x) + threadIdx.x;
+  if (i >= n) return;
+  row_sizes[i] += (row_sizes2[i] + k);
+}
+
+/**
+ * @brief Perform data + data.T operation.
+ * Can only run once row_sizes from the CSR matrix of data + data.T has been
+ * determined.
+ *
+ * @param edges: Input row sum array(n) after reduction
+ * @param data: Input knn distances(n, k)
+ * @param indices: Input knn indices(n, k)
+ * @param VAL: Output values for data + data.T
+ * @param COL: Output column indices for data + data.T
+ * @param ROW: Output row indices for data + data.T
+ * @param n: Number of rows
+ * @param k: Number of n_neighbors
+ */
+template <typename math_t>
+__global__ static void symmetric_sum(int *restrict edges,
+                                     const math_t *restrict data,
+                                     const long *restrict indices,
+                                     math_t *restrict VAL, int *restrict COL,
+                                     int *restrict ROW, const int n,
+                                     const int k) {
+  const int j =
+    (blockIdx.x * blockDim.x) + threadIdx.x;  // for every item in row
+  const int row = (blockIdx.y * blockDim.y) + threadIdx.y;  // for every row
+  if (row >= n || j >= k) return;
+
+  const int col = indices[row * k + j];
+  const int original = atomicAdd(&edges[row], 1);
+  const int transpose = atomicAdd(&edges[col], 1);
+
+  VAL[transpose] = VAL[original] = data[row * k + j];
+  // Notice swapped ROW, COL since transpose
+  ROW[original] = row;
+  COL[original] = col;
+
+  ROW[transpose] = col;
+  COL[transpose] = row;
+}
+
+/**
+ * @brief Perform data + data.T on raw KNN data.
+ * The following steps are invoked:
+ * (1) Find how much space needed in each row
+ * (2) Compute final space needed (n*k + sum(row_sizes)) == 2*n*k
+ * (3) Allocate new space
+ * (4) Prepare edges for each new row
+ * (5) Perform final data + data.T operation
+ * (6) Return summed up VAL, COL, ROW
+ *
+ * @param knn_indices: Input knn distances(n, k)
+ * @param knn_dists: Input knn indices(n, k)
+ * @param n: Number of rows
+ * @param k: Number of n_neighbors
+ * @param out: Output COO Matrix class
+ * @param stream: Input cuda stream
+ */
+template <typename math_t, int TPB_X = 32, int TPB_Y = 32>
+void from_knn_symmetrize_matrix(const long *restrict knn_indices,
```

Comment:
This should accept a `shared_ptr<deviceAllocator>` (which you can pull from the cumlHandle when calling it).